doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+Then things like `mmapstorage` could be covered as well.
+
+```c
+// Based on: https://codereview.stackexchange.com/questions/146561/reference-counting-in-c99
+
+// Note: should NOT call free() on pointer
+typedef void (*Destructor)(const void*);
+
+typedef struct {
+  Destructor destroy;
+  unsigned int count;
+} ObjectRecord;
+
+static ObjectRecord *objectToRecord(void *object) { return (ObjectRecord*)(object) - 1; }
+static void *recordToObject(ObjectRecord *record) { return record + 1; }
+
+void * elektraMalloc(size_t size, Destructor destroy) {
+  ObjectRecord *record = malloc (sizeof (ObjectRecord) + size);
+  if (record == NULL) {
+    return NULL;
+  }
+  record->count = 1;
+  record->destroy = destroy;
+  return recordToObject (record);
+}
+
+unsigned int elektraRefCount(const void * p) {
+  return objectToRecord (p)->count;
+}
+
+void elektraRetain(const void *p)
+{
+    ++objectToRecord(p)->count;
+}
+
+void elektraRelease(const void *p)
+{
+    ObjectRecord *record = objectToRecord(p);
+
+    record->count--;
+    if (record->count > 0) {
+        /* still referenced */
+        return;
+    }
+
+    if (record->destory) {
+        record->destory(p);
+    }
+    free(record);
+}
+```
+
+Usage example:
+
+```c
+static void ksDestroy(const void * p) {
+  KeySet * ks = (KeySet *)p;
+  for (size_t i = 0; i < ks->size; i++) {
+    elektraRelease(key);
+  }
+  // [...]
+}
+
+KeySet * ksCreate() {
+  KeySet * ks = elektraMalloc(sizeof(struct KeySet), ksDestroy);
+  // [...] initalization
+  return ks;
+}
+
+void ksAppendKey(KeySet * ks, Key * key) {
+  elektraRetain(key);
+  // [...]
+}
+
+// for compatibility and for complete API
+static inline void ksDel(KeySet * ks) {
+  elektraRelease(ks);
+}
+```
+
+## Decision
+
+General decisions regardless of implementation:
+
+- `Key` has two counters.
+  One that blocks deletion, another to prevent name changes.
+- Reference counts are of type `uint16_t`.
+
+### Counted deletion lock
+
+The reference counting consists of two functions `inc` and `dec`.
+When a structure is created, its reference count is set to `0`.
+Only in this state can the structure be deleted.
+If the count is non-zero, deletion is no-op.
+The `inc` and `dec` function increment and decrement this count.
+
+Therefore, the reference counter can be considered a reentrant/counting lock, which blocks the deletion of a structure.
+At zero the lock is unlocked, any calls to `inc` lock the lock one more time, and calls to `dec` unlock once.
+
+The secondary counter for the name of a `Key` works the same way, but it blocks changes to the name.
+
+Additional details:
+
+- `inc` returns an error on reference count overflow
+- `dec` does not return an error on underflow, instead the count stays at zero
+
+## Rationale
+
+- Adding a second reference counter to Key and reducing the size of both significantly (`size_t` to `uint16_t`)
+  actually saves memory (32 vs 64bit on 64-bit machines) compared to the previous solution.
```

Comment:
This is actually a left over bit from the old version of this decision. I'll probably split out the stuff about locking the key name into a separate decision. Maybe it'll end up in #4689. The second counter looks like our current ref counter, but it really has a completely different use case.

> it's not clear without implementation detail whether this really saves memory.

In the old version the decision stated that we go from one `size_t` to two `uint16_t` right next to each other in `struct _Key`. In that case it could save memory or stay the same, but it definitely wouldn't need more.